### PR TITLE
Improve local reference point mapping suggestion

### PR DESCRIPTION
**Pull request by tobous**
Improves the logic suggesting existing directories for an incoming resource negotiation to work for arbitrary reference point base resources. This is done by sending proposed representation options as part of the additional resource negotiation data that can be used to determine the best mapping on the receiving side.

### Reviewing this PR

I would suggest reviewing this PR commit by commit.

### Commits

<details><summary><b>[FEATURE][CORE] Provide possible reference point representation options</b></summary>
<br>

Extends the AdditionalResourceData by adding possible resource
representations. These possible representation options are defined
through the name of a reference object (e.g. Eclipse project, IntelliJ
module, etc.) and the path of the reference point resource relative to
that object. This should hopefully allow the receiving side to make a
good suggestion on how to map the received reference point if its
representation already exists on the receiving side.

A list of such representations is offered as the representation does not
always have to be unique.

The class AbstractPossibleRepresentationProvider can be extended by the
specific IDE implementations which have to provide the logic to obtain
the possible representations for a given reference point.

The data can be read from the resource negotiation data using a utility
method of AbstractPossibleRepresentationProvider. This was added as the
data has to be flattened and escaped multiple times before adding it to
the additional resource data mapping, so the reversal of that process is
also done by the class.

</details>

<details><summary><b>[FEATURE][E] Add EclipsePossibleRepresentationProvider</b></summary>
<br>

The provider uses the project name and project-relative path of the
reference point resource as the possible representation.

</details>

<details><summary><b>[FEATURE][I] Add IntellijPossibleRepresentationProvider</b></summary>
<br>

The provider uses the module name and content-root-relative path of all
modules the reference point resource is contained in (even if only in a
nested fashion) as the possible representations.

The provider explicitly adds all possible representations as the
IntelliJ module system allows for more intricate (nested) setups that
are not necessarily representable in all other IDE resource models (i.e.
in the Eclipse project model). As a result, the nearest parent module
might not always be the right reference object for the receiving side.

As an example, when sharing a Gradle project, we are interested in the
second parent module – as Gradle creates separate child modules for the
main and test sources in IntelliJ – instead of the direct parent module.

Makes the method IntellijReferencePoint.getVirtualFile() public to make
it accessible to the representation provider implementation.

</details>

<details><summary><b>[FEATURE][E] Use possible representation suggestions</b></summary>
<br>

Uses the possible representation suggestions to suggest an initial
reference point mapping for the incoming resource negotiation.

</details>

<details><summary><b>[FEATURE][I] Use possible representation suggestions</b></summary>
<br>

Uses the possible representation suggestions to suggest an initial
reference point mapping for the incoming resource negotiation.

</details>

### CODE REVIEW

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java  2
- eclipse/src/saros/negotiation/additional_resource_data/EclipsePossibleRepresentationProvider.java  1
- intellij/src/saros/negotiation/additional_resource_data/IntellijPossibleRepresentationProvider.java  4
- eclipse/src/saros/ui/wizards/pages/LocalRepresentationSelectionPage.java  3
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/ui/wizards/pages/referencepointselection/ReferencePointTab.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5767479&bid=18875925)

File: core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+                path = "";
+              }
+
+              return new ImmutablePair<>(ESCAPE_INNER.unescape(name), ESCAPE_INNER.unescape(path));
+            })
+        .collect(Collectors.toList());
+  }
+
+  /** Utility class to escape, unescape, and split strings. */
+  private static class EscapeUtils {
```
Comment (srossbach) [inline core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java:138]: This is somewhat generic. I already used this when remembering project selections. Can you please move it to the util package ?

File: eclipse/src/saros/negotiation/additional_resource_data/EclipsePossibleRepresentationProvider.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    super(additionalResourceDataFactory);
+  }
+
+  @Override
+  protected List<Pair<String, String>> computePossibleRepresentations(
+      IReferencePoint referencePoint) {
+    IContainer referencePointDelegate = ResourceConverter.getDelegate(referencePoint);
+
+    String projectName = referencePointDelegate.getProject().getName();
+    String projectRelativePath = referencePointDelegate.getProjectRelativePath().toPortableString();
```
Comment (srossbach) [inline eclipse/src/saros/negotiation/additional_resource_data/EclipsePossibleRepresentationProvider.java:30]: This is Eclipse specific (the toPortableString() call). I do not see any hints from which IDE this data comes from. Maybe the method does what you need but it is not specified anywhere how a relative should look like.

Can I assume that  the directory Foo\Bar in a Unix environment is correctly invalidated on a Windows machine ?

**Comment by tobous** [comment]
Rebased onto current base branch without any interaction.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- eclipse/src/saros/ui/wizards/pages/LocalRepresentationSelectionPage.java  3
- intellij/src/saros/negotiation/additional_resource_data/IntellijPossibleRepresentationProvider.java  4
- core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java  2
- eclipse/src/saros/negotiation/additional_resource_data/EclipsePossibleRepresentationProvider.java  1
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/ui/wizards/pages/referencepointselection/ReferencePointTab.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5767479&bid=18875925)

File: core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java (hunk 1 above)
Comment (tobous) [reply core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java]: Done. Moved to class EscapeUtil.

File: eclipse/src/saros/negotiation/additional_resource_data/EclipsePossibleRepresentationProvider.java (hunk 2 above)
Comment (tobous) [reply eclipse/src/saros/negotiation/additional_resource_data/EclipsePossibleRepresentationProvider.java:30]: Added javadoc stating that the paths have to use Unix separators. Also adjusts the logic on the receiving side for Eclipse to use `Path.fromPortableString()` to correctly parse the Unix-separated path on Windows.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- intellij/src/saros/negotiation/additional_resource_data/IntellijPossibleRepresentationProvider.java  4
- core/src/saros/util/EscapeUtil.java  1
- eclipse/src/saros/ui/wizards/pages/LocalRepresentationSelectionPage.java  3
- eclipse/src/saros/negotiation/additional_resource_data/EclipsePossibleRepresentationProvider.java  1
- core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java  2
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/ui/wizards/pages/referencepointselection/ReferencePointTab.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5767479&bid=18875925)

**Comment by tobous** [comment]
Rebased onto current base branch without any interaction.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- core/src/saros/util/EscapeUtil.java  1
- eclipse/src/saros/negotiation/additional_resource_data/EclipsePossibleRepresentationProvider.java  1
- core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java  2
- intellij/src/saros/negotiation/additional_resource_data/IntellijPossibleRepresentationProvider.java  4
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/ui/wizards/pages/referencepointselection/ReferencePointTab.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5767479&bid=18875925)

**Comment by tobous** [comment]
Rebased onto current base branch without any interactions.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- intellij/src/saros/negotiation/additional_resource_data/IntellijPossibleRepresentationProvider.java  4
- core/src/saros/util/EscapeUtil.java  1
- core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java  2
- eclipse/src/saros/negotiation/additional_resource_data/EclipsePossibleRepresentationProvider.java  1
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/ui/wizards/pages/referencepointselection/ReferencePointTab.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5767479&bid=18875925)

**Comment by tobous** [comment]
Rebased onto current base branch without any interaction.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- core/src/saros/util/EscapeUtil.java  1
- core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java  2
- intellij/src/saros/negotiation/additional_resource_data/IntellijPossibleRepresentationProvider.java  4
- eclipse/src/saros/negotiation/additional_resource_data/EclipsePossibleRepresentationProvider.java  1
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/ui/wizards/pages/referencepointselection/ReferencePointTab.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5767479&bid=18875925)

**Comment by tobous** [comment]
Rebased onto current master without any interactions.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- eclipse/src/saros/negotiation/additional_resource_data/EclipsePossibleRepresentationProvider.java  1
- intellij/src/saros/negotiation/additional_resource_data/IntellijPossibleRepresentationProvider.java  4
- core/src/saros/util/EscapeUtil.java  1
- core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java  2
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/ui/wizards/pages/referencepointselection/ReferencePointTab.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5767479&bid=18875925)

**Comment by tobous** [comment]
@srossbach @m273d15 Almost done. I would still like to get this PR and #1058 merged before the end of October. Could you review the changes? This PR looks like a lot but it is mostly just boilerplate stuff necessary to handle the sent values.

File: core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+   * @return the possible representations contained in the given string
+   */
+  private static List<Pair<String, String>> split(String flattenedPossibleRepresentations) {
+    return ESCAPE_OUTER
+        .getSplitPattern()
+        .splitAsStream(flattenedPossibleRepresentations)
+        .map(ESCAPE_OUTER::unescape)
+        .map(
+            joinedPossibleRepresentation -> {
+              String[] values = ESCAPE_INNER.getSplitPattern().split(joinedPossibleRepresentation);
```
Comment (srossbach) [inline core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java:115]: What about moving the code here to its own method so the function statement is easier to read ?

File: core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+              if (values.length != 1 && values.length != 2) {
+                throw new IllegalStateException(
```
Comment (srossbach) [inline core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java:118]: I guess this exception is not catched anywhere and I guess this is always called from some UI code. In other words bad data can make the UI non useable which is not very good. Isn't is better just to discard the entries or simply not return anything ?

File: core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java (hunk 3 above)
Comment (tobous) [reply core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java]: Done.

File: core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java (hunk 4 above)
Comment (tobous) [reply core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java]: Done. Replaced with error log.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java  2
- core/src/saros/util/EscapeUtil.java  1
- intellij/src/saros/negotiation/additional_resource_data/IntellijPossibleRepresentationProvider.java  4
- eclipse/src/saros/negotiation/additional_resource_data/EclipsePossibleRepresentationProvider.java  1
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/ui/wizards/pages/referencepointselection/ReferencePointTab.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5767479&bid=18875925)

File: core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java (hunk 3 above)
Comment (tobous) [reply core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java]: Moved the split process to a separate method as well.

Comment (srossbach) [reply core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java]: My bad. I meant the whole process so it reads like

``` .map(ESCAPE_OUTER::unescape).map(someMeaningfullName).filter(...).collect()

Comment (tobous) [reply core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java]: > My bad. I meant the whole process so it reads like
> 
> `.map(ESCAPE_OUTER::unescape).map(someMeaningfullName).filter(...).collect()`

Timing^^ already done.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- core/src/saros/negotiation/additional_resource_data/AbstractPossibleRepresentationProvider.java  4
- intellij/src/saros/negotiation/additional_resource_data/IntellijPossibleRepresentationProvider.java  4
- eclipse/src/saros/negotiation/additional_resource_data/EclipsePossibleRepresentationProvider.java  1
- core/src/saros/util/EscapeUtil.java  1
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/ui/wizards/pages/referencepointselection/ReferencePointTab.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5767479&bid=18875925)

**Review by srossbach — APPROVED** [review]

**Pull request merged**